savefile/savefile.go

```diff
@@ -0,0 +1,129 @@
+package savefile
+
+import (
+	"bytes"
+	"compress/zlib"
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/sector-f/jhmod/nvc"
+)
+
+const (
+	// Magic header for Jupiter Hell save save files.
+	magic = "\xde\xc0\xad\xde"
+)
+
+type savedata struct {
+	// Player name
+	PlayerName string
+	// Game mode.  This can be "jh", and various others.
+	GameMode string
+	// The current level's name.
+	CurrentLevel string
+	// The seed used to generate the game.
+	Seed uint32
+}
+
+// Something was reading for data but got an EOF.
+var UnexpectedEndOfFileErr error = errors.New("unexpected eof")
+
+// General parse error.  TODO make more descriptive errors.
+var SaveParseErr error = errors.New("general parse error")
+
+// Read in a string from a buffer of size suffixed with a length byte.
+func readString(r io.Reader, size int) (string, error) {
+	buf := make([]byte, size)
+	nRead, err := r.Read(buf[:])
+	if err != nil {
+		return "", err
+	}
+	if nRead != size {
+		return "", UnexpectedEndOfFileErr
+	}
```

Comment:
This check should be unnecessary if you switch to `io.ReadFull`:

> On return, n == len(buf) if and only if err == nil.

And `size` will equal `len(buf)`, because `buf` was created with `size` as its length, meaning that if `nRead != size` then you would've already returned because `err` was non-nil

Reply:
:boom: